apps/btc_app/get_public_key.c

```diff
@@ -0,0 +1,292 @@
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->get_public_key.which_request) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request_data(btc_get_public_key_request_t *request) {
+  bool status = true;
+
+  // TODO: Enable btc/coin specific check
+  if (false ==
+      verify_receive_derivation_path(request->initiate.derivation_path,
+                                     request->initiate.derivation_path_count)) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+  return status;
+}
+
+static size_t btc_get_address(const uint8_t *seed,
+                              const uint32_t *path,
+                              uint32_t path_length,
+                              uint8_t *public_key,
+                              char *address) {
+  HDNode node = {0};
+  uint8_t address_version = 0;
+  uint32_t purpose_index = path[0];
+  uint32_t coin_index = path[1];
+  char addr[50] = "";
+  size_t address_length = 0;
+
+  if (!derive_hdnode_from_path(
+          path, path_length, SECP256K1_NAME, seed, &node)) {
+    // send unknown error; unknown failure reason
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+    memzero(&node, sizeof(HDNode));
+    return address_length;
+  }
+
+  switch (purpose_index) {
+    case NATIVE_SEGWIT:
+      // ignoring the return status and handling by size of address
+      get_segwit_address(
+          node.public_key, sizeof(node.public_key), coin_index, addr);
+      break;
+    case NON_SEGWIT:
+      get_version(purpose_index, coin_index, &address_version, NULL);
+      hdnode_get_address(&node, address_version, addr, 35);
+      break;
+    // TODO: add support for taproot and segwit
+    default:
+      break;
+  }
+
+  address_length = strnlen(addr, sizeof(addr));
+  if (0 >= address_length) {
+    // address encoding failed
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+  }
+  if (NULL != public_key) {
+    ecdsa_uncompress_pubkey(
+        get_curve_by_name(SECP256K1_NAME)->params, node.public_key, public_key);
+  }
+  if (NULL != address) {
+    memcpy(address, addr, address_length);
+  }
+  memzero(&node, sizeof(HDNode));
+  return address_length;
+}
+
+static void send_public_key(const uint8_t *public_key) {
+  btc_result_t response = init_btc_result(BTC_RESULT_GET_PUBLIC_KEY_TAG);
+  response.get_public_key.which_response =
+      BTC_GET_PUBLIC_KEY_RESPONSE_RESULT_TAG;
+  ecdsa_uncompress_pubkey(get_curve_by_name(SECP256K1_NAME)->params,
+                          public_key,
+                          response.get_public_key.result.public_key);
```

Comment:
Should uncompress only once, same handling added in `btc_get_address` as well.

Reply:
changed